lib/cn/cn_tree.c

```diff
@@ -1405,10 +1406,18 @@ cn_comp_release(struct cn_compaction_work *w)
 
     perfc_inc(w->cw_pc, PERFC_BA_CNCOMP_FINISH);
 
-    if (w->cw_completion)
-        w->cw_completion(w);
-    else
+    if (HSE_UNLIKELY(!w->cw_completion)) {
         free(w);
+        return;
+    }
+
+    sts_job_status_set(&w->cw_job, 'Z');
+
+    /* After this function returns the job will be disassociated
+     * from its thread and hence becomes a zombie.  Do not touch
+     * *w afterward as it may have already been freed.
+     */
+    w->cw_completion(w);
```

Comment:
If only there was a way to tell the compiler, "don't allow usage after this line"

Reply:
w = NULL;